jlm/llvm/opt/ScalarEvolution.cpp

```diff
@@ -54,28 +139,45 @@ ScalarEvolution::Run(
   auto statistics = Statistics::Create(rvsdgModule.SourceFilePath().value());
   statistics->Start();
 
+  Context_ = Context::Create();
   InductionVariableMap_.clear();
   const rvsdg::Region & rootRegion = rvsdgModule.Rvsdg().GetRootRegion();
-  TraverseRegion(rootRegion);
+  AnalyzeRegion(rootRegion);
 
-  statistics->Stop();
+  statistics->Stop(*Context_.get());
   statisticsCollector.CollectDemandedStatistics(std::move(statistics));
 };
 
 void
-ScalarEvolution::TraverseRegion(const rvsdg::Region & region)
+ScalarEvolution::AnalyzeRegion(const rvsdg::Region & region)
 {
   for (const auto & node : region.Nodes())
   {
     if (const auto structuralNode = dynamic_cast<const rvsdg::StructuralNode *>(&node))
     {
       for (auto & subregion : structuralNode->Subregions())
       {
-        TraverseRegion(subregion);
+        AnalyzeRegion(subregion);
       }
       if (const auto thetaNode = dynamic_cast<const rvsdg::ThetaNode *>(structuralNode))
       {
-        PerformSCEVAnalysis(*thetaNode);
+        // Add number of loop vars in theta (for statistics)
+        for (const auto loopVar : thetaNode->GetLoopVars())
+        {
+          Context_.get()->AddLoopVar(*loopVar.pre);
+        }
+
+        auto chrecMap = PerformSCEVAnalysis(*thetaNode);
+        for (auto & [output, chrec] : chrecMap)
+        {
+          // Add computed (non-trivially unknown) chrecs to context for statistics
+          const bool isTrivialUnknown =
```

Comment:
Could this logic be moved to a method on the class called `isTrivialUnknown` or maybe just `isUnknown`, with a documentation string that explains what it means for a Chain Recurrence to be unknown?

If a chain recurrence contains any step that is unknown, does it not make sense to regard the whole thing as unknown?